contracts/token/ERC721/ERC721Enumerable.sol

```diff
@@ -144,4 +147,46 @@ contract ERC721Enumerable is ERC165, ERC721, IERC721Enumerable {
         _allTokensIndex[tokenId] = 0;
         _allTokensIndex[lastToken] = tokenIndex;
     }
+
+    /**
+     * @dev Private function to add a token to this extension's ownership-tracking data structures.
+     * @param to address representing the new owner of the given token ID
+     * @param tokenId uint256 ID of the token to be added to the tokens list of the given address
+     */
+    function _addTokenToOwnerEnumeration(address to, uint256 tokenId) private {
+        uint256 newOwnedTokensLength = _ownedTokens[to].push(tokenId);
+        // No need to use SafeMath since the length after a push cannot be zero
+        _ownedTokensIndex[tokenId] = newOwnedTokensLength - 1;
+    }
+
+    /**
+     * @dev Private function to remove a token from this extension's ownership-tracking data structures. Note that
+     * while the token is not assigned a new owner, the _ownedTokensIndex mapping is _not_ updated: this allows for
+     * gas optimizations e.g. when performing a transfer operation (avoiding double writes).
+     * This has O(1) time complexity, but alters the order of the _ownedTokens array.
+     * @param from address representing the previous owner of the given token ID
+     * @param tokenId uint256 ID of the token to be removed from the tokens list of the given address
+     */
+    function _removeTokenFromOwnerEnumeration(address from, uint256 tokenId) private {
+        // To prevent a gap in from's tokens array, we store the last token in the index of the token to delete, and
+        // then delete the last slot (swap and pop).
+
+        uint256 lastTokenIndex = _ownedTokens[from].length.sub(1);
+        uint256 lastTokenId = _ownedTokens[from][lastTokenIndex];
+
+        uint256 tokenIndex = _ownedTokensIndex[tokenId];
+
+        _ownedTokens[from][tokenIndex] = lastTokenId; // Move the last token to the slot of the to-delete token
+        _ownedTokensIndex[lastTokenId] = tokenIndex; // Update the moved token's index
+
+        // Note that this will handle single-element arrays. In that case, both tokenIndex and lastTokenIndex are going
+        // to be zero. The swap operation will therefore have no effect, but the token _will_ be deleted during the
+        // 'pop' operation.
+
+        // This also deletes the contents at the last position of the array
+        _ownedTokens[from].length--;
+
+        // Note that _ownedTokensIndex[tokenId] hasn't been cleared: it still points to the old slot (now occcupied by
```

Comment:
I'm not a fan of this tbh. It's not that bad because it's `private` data, and the interface exposed is clean. But it feels like an error waiting to happen. I always expect "unused" slots in mappings to be zeroed out, and I'm sure most Solidity programmers expect the same.

We should at least have a comment at the place where the state variable is defined saying that it should not be assumed clean.

Reply:
An error waiting to happen during OZ development? I'd argue the peer-review process and the test suite should prevent said errors, and would trade this internal complexity for a cleaner API with reduced gas costs.

That said, as described [here](https://github.com/OpenZeppelin/openzeppelin-solidity/pull/1539#discussion_r239599345) I expect this to be just the first step in a 721 internals refactor, I'm also not thrilled with the state this PR leaves things in (though I prefer that to not fixing the issue).

Reply:
There's also the problem of upgradeable ERC721 instances. How can we ensure that a future version won't assume the mapping slot to be zeroed?

Reply:
Because zero is a valid index value, not a sentinel one. We _could_ migrate to 1-based indexing and `require(index != 0)` e.g. when removing, but that's a whole different topic.

Reply:
Additionally, unused values of the mapping _will_ be zeroed-out: calls to `transfer` don't clear the slot because it is about to be reassigned (in `_add`).

Reply:
Which internals, `_transfer`? There's a missing test case, though a complete suite refactor is due imo (#1148). I can add it to this PR to keep full coverage.

Reply:
I meant `remove`.